[UI Tests] Rolling Balance + Comprehensive UI Testing

sp4ce-cowboy opened the pull request:
### Overview
- UserInputBuilder class has been added for `FxRobot` to be able to emulate user input
- Rolling Balance Indicator now has its own UITest class with code refactored as much as possible

### Classes Affected (4)
- No change to functional code, all additions are isolated to test package only
- RollingBalanceUiTest added, UserInputBuilder added, TestDataUtil added
- Removed the old Rolling Balance test from MainAppUiTest

https://github.com/AY2324S1-CS2103-T16-3/tp/assets/19762596/4f80c537-a33a-4b13-92da-b03855fefae6

### Other considerations
- While intended for Rolling Balance Indicator primarily, the test is comprehensive enough to now be available for other elements as well, since `FxRobot` can basically emulate any user input incl. valid/invalid transaction/error input/etc. We can essentially make a whole product demo with all commands, errors, results, etc with just `FxRobot`
- If you are doing UI tests locally, it might take a while to complete (~5 minutes)

woojiahao commented on src/test/java/unicash/ui/RollingBalanceUiTest.java:
```diff
@@ -0,0 +1,179 @@
+    @BeforeEach
+    public void setUp() throws TimeoutException {
+        FxToolkit.registerPrimaryStage();
+        FxToolkit.setupApplication(() -> new MainApp());
```
Please use the `MainApp(Path)` variant instead. Using just `MainApp()` will cause your unit tests to override the actual data file on your machine and might cause problems with the CI if that data is needed. See this for reference: https://ay2324s1-cs2103-t16-3.github.io/tp/DeveloperGuide.html#using-beforeeach

sp4ce-cowboy replied:
Thanks for letting me know! I have added the change.

woojiahao reviewed (changes requested):
Comment about using the `MainApp(Path)` constructor. Just curious but are you planning to continue adding more unit testing? I think at this stage, we're good with the amount of coverage we currently have is one of the, if not the, highest coverage so we can avoid touching the code already

sp4ce-cowboy commented:
> Comment about using the `MainApp(Path)` constructor. Just curious but are you planning to continue adding more unit testing? I think at this stage, we're good with the amount of coverage we currently have is one of the, if not the, highest coverage so we can avoid touching the code already

@woojiahao Thanks for letting me know about the issue with the constructor, I have made the change!

Regarding the unit testing, I believe these are the last tests that I'm going to add for the UI components that I have implemented. I agree that we have good coverage, however this is not particularly about overall coverage, but rather the 3A Developer Testing tp deliverable requirement  (_..."Evaulates: how well you tested your own feature")_  and I felt that the RollingBalance Indicator was not sufficiently tested, thus I have added the tests to avoid any possible penalty. No other files were affected and no authorship of code was affected in the process!

woojiahao commented:
Yeap no issues with the unit tests, just thinking we should wrap up code changes unless urgent

woojiahao reviewed (dismissed):
Just checking but you might want to go through the rest of your UI tests to make sure the filepath issue isn't present there too

sp4ce-cowboy commented:
> Just checking but you might want to go through the rest of your UI tests to make sure the filepath issue isn't present there too

Alls good no worries, none of the other UI Tests require MainApp. Double checked `tests.ui` just now and everything looks good!

sp4ce-cowboy commented:
> Yeap no issues with the unit tests, just thinking we should wrap up code changes unless urgent

Yup I agree! I think all other UI Components (Data Source Indicator, CommandBox, SummaryWindow, HelpWindow, StyleSheet, and MainApp) have UI tests so Rolling Balance should be the last and final UI component.

woojiahao reviewed (approved):
Cool, great work, LGTM
